### ISSUE TITLE
Feature request: nonblocking calls to S3

### ISSUE THREAD

**Issue by JohnHBrock** [amitmurthy/AWS.jl#39]
It would be nice to have the various calls to S3 (e.g., get_object) optionally be nonblocking in order to facilitate asynchronous IO.

I was hoping it would be as simple as setting RequestOptions.blocking to false before calling HTTPC.get(), but that results in a conversion error in AWS.S3.do_request: for reasons I don't understand, the header's Date field apparently contains an Array instead of a String when RequestOptions.blocking == false. I can't tell if that's a bug or a feature.

**Comment by s2maki** [comment]
We were working on this as well, but weren't able to reach a satisfying conclusion for our particular use case (hundreds of open S3 connections streaming simultaneously).  Our main problem was that when we left a socket idle for a few seconds, S3 would close it due to inactivity.  But we did have code that worked as long as you were reading frequently enough.  I'll look to see if it is production suitable for release.

For our particular use case, we ended up using the Range header to read smaller chunks of data from each S3 key.  We'd read 5MB at a time and hold it in a buffer to use as we needed it, then pull the next 5MB when the buffer ran low.  This isn't an ideal solution, but it might be a good enough workaround depending on what you're trying to do.

**Comment by s2maki** [comment]
Oh, and by "we", I mean an independent group.  I'm not actually affiliated with the maintainer of the library, just an occasional contributor.

**Comment by s2maki** [comment]
I just checked, and unfortunately our streaming code is not sufficiently complete to release.  We were doing work to read multiple sockets at once with a single curl_multi handle.  But here's some info I can share that may shed a little light on the work that needs to be done to support streaming a single socket.

AWS.jl uses HTTPClient, which in turn is a wrapper for LibCURL.jl.  HTTPClient doesn't appear to directly support streaming.  But it can dump data to a file or write it into a subtype of IO.  In theory, you ought to be able to create a concrete subtype of IO that has a memory based buffer similar to PipeBuffer, except that it would have a limited size and would block writes to it when the buffer is full. You could call get_object asynchronously (in a task, with @async) and pass an instance of this blocking pipebuffer as its fifth argument.  Your reader would read from this IO instance in your normal thread.  You would have to implement EOF handling as well, to have the async task exit and signal your reader that there was no more data.

I think this all could be done without any special support from AWS.jl or HTTPClient.jl.

I hope this helps.

**Comment by JohnHBrock** [comment]
Thanks, that's all interesting and helpful. I may indeed try to solve my particular problem by bypassing AWS.jl and HTTPClient.jl altogether.

**Comment by sebastiang** [comment]
I would love to hear if any work is done in this regard

**Comment by amitmurthy** [comment]
Wouldn't this work for you?

```
  t = @async S3.get_object(.....)
......   # other code
  result = wait(t)
```

The wait() call will block and result the result of `get_object`

Depending on your use case you can also have the `@async` block write out the result to a `Channel` or trigger a `Condition` variable and wait on either on either of them.

**Comment by sebastiang** [comment]
Yes, you're probably right!

**Comment by JohnHBrock** [comment]
Will `@async` allow for truly simultaneous IO when used with `S3.get_object(...)`? I'm thinking specifically of the case where I'd want dozens of downloads from S3 simultaneously.

**Comment by amitmurthy** [comment]
Yes it will.

**Comment by s2maki** [comment]
Correct me if I'm wrong, but `S3.get_object` will still read the entire object into memory before being able to return anything.  So if you're streaming dozens of very large objects direct to RAM, you're going to need a lot of RAM.  If all of your objects are small enough to fit into memory in their entirety, then it doesn't matter much.

For larger objects, `get_object` does have the ability to take an instance of IO type, which could be used to throttle the reading.  For example, you could have `get_object` write the data out to disk and then read from the disk file at your leisure.  In order to make that transparent to higher level code (for example, a function that emulated the behavior of `open`/`read` on an S3 object as though it were a local file), a custom IO sub-type would have to be implemented.  I've tried my hand at this, but have only achieved a partially satisfactory result.

Additionally, each request to `get_object` causes HTTPClient to create a new CURL multi handle, which incurs a significant amount of overhead for each new connection.  This overhead is very noticeable when opening and closing a lot of S3 sockets.  I have some experimental code that pools curlm instances so that the CURL easy handle's socket remain open for the next call.  But it still operates on a per-task basis.  I have not yet been able to have a single curlm across multiple asynchronous tasks while returning each socket's data to the correct task.

**Comment by amitmurthy** [comment]
Right. Migrating to Requests.jl may be the right thing to do for the CURL multi handle issue.

**Comment by sebastiang** [comment]
I cracked open the code to see how hard this would be. I’ll tell you in an hour...
On November 23, 2015 at 9:25:10 PM, Amit Murthy (notifications@github.com) wrote:

Right. Migrating to Requests.jl may be the right thing to do for the CURL multi handle issue.

**Comment by samoconnor** [comment]
Note that AWSS3.jl has an option to return an IO stream instead of buffering the entire result into memory. See example here: https://github.com/samoconnor/AWSS3.jl/blob/master/src/AWSS3.jl#L112

Combining this with @async should allow efficient simultaneous processing of multiple downloads.
